File: UnitTests/OPStest/UnitTests/SubscribeDataAndTest.cpp
// PizzaTest.cpp : Defines the entry point for the console application.
//
#include "SubscribeDataAndTest.h"
#include "gtest/gtest.h"
#include "TestPizza.cpp"



#undef USE_MESSAGE_HEADER

#ifndef _WIN32
#include <time.h>

__int64 getNow()
{
    struct timespec ts;
    memset(&ts, 0, sizeof(ts));
    //clock_gettime(CLOCK_REALTIME, &ts);
    return ((1000 * ts.tv_sec) + (ts.tv_nsec / 1000000));
}

#include <stdio.h>
#include <sys/select.h>
#include <sys/ioctl.h>
#include <termios.h>

int _kbhit() {
    static const int STDIN = 0;
    static bool initialized = false;

    if (! initialized) {
        // Use termios to turn off line buffering
        termios term;
        tcgetattr(STDIN, &term);
        term.c_lflag &= ~ICANON;
        tcsetattr(STDIN, TCSANOW, &term);
        setbuf(stdin, NULL);
        initialized = true;
    }

    int bytesWaiting;
    ioctl(STDIN, FIONREAD, &bytesWaiting);
    return bytesWaiting;
}
#else
__int64 getNow()
{
    return (__int64)timeGetTime();
}
#endif

template <class DataType>
class CHelperListener
{
public:
	virtual void onData(ops::Subscriber* sub, DataType* data) = 0;
};

class IHelper
{
public:
	virtual void CreateSubscriber(ops::Participant* part, std::string topicName) = 0;
	virtual void DeleteSubscriber(bool doLog = true) = 0;
	virtual void StartSubscriber() = 0;
	virtual void StopSubscriber() = 0;
	virtual ~IHelper() {};
};

template <class DataType, class DataTypePublisher, class DataTypeSubscriber>
class CHelper : public IHelper, ops::DataListener, ops::DeadlineMissedListener
{
private:
	CHelperListener<DataType>* client;
	ops::Publisher* pub;
	ops::Subscriber* sub;
	__int64 expectedPubId;

public:
	DataType data;

	CHelper(CHelperListener<DataType>* client):
		pub(NULL), sub(NULL), expectedPubId(-1)
	{
		this->client = client;
	}

	virtual ~CHelper()
	{
		DeleteSubscriber(false);
	}



	void CreateSubscriber(ops::Participant* part, std::string topicName)
	{
		if (sub) {
			std::cout << "Subscriber already exist for topic " << sub->getTopic().getName() << std::endl;
		} else {
			try {
				//Create topic, might throw ops::NoSuchTopicException
				ops::Topic topic = part->createTopic(topicName);

				//Create a subscriber on that topic.
				sub = new DataTypeSubscriber(topic);
				sub->addDataListener(this);
				sub->deadlineMissedEvent.addDeadlineMissedListener(this);

				sub->start();
			}
			catch (...) {
				std::cout << "Requested topic '" << topicName << "' does not exist!!" << std::endl;
			}
		}
	}

	void DeleteSubscriber(bool doLog = true)
	{
		if (sub) {
			std::cout << "Deleting subscriber for topic " << sub->getTopic().getName() << std::endl;
			sub->stop();
			delete sub;
			sub = NULL;
		} else {
			if (doLog) std::cout << "Subscriber must be created first!!" << std::endl;
		}
	}

	void StartSubscriber()
	{
		if (sub) {
			std::cout << "Starting subscriber for topic " << sub->getTopic().getName() << std::endl;
			sub->start();
		} else {
			std::cout << "Subscriber must be created first!!" << std::endl;
		}

	}

	void StopSubscriber()
	{
		if (sub) {
			std::cout << "Stoping subscriber for topic " << sub->getTopic().getName() << std::endl;
			sub->stop();
		} else {
			std::cout << "Subscriber must be created first!!" << std::endl;
		}
	}
	void SetDeadlineQos(__int64 timeoutMs)
	{
		if (sub) {
			std::cout << "Setting deadlineQos to " << timeoutMs << " [ms] for topic " << sub->getTopic().getName() << std::endl;
			sub->setDeadlineQoS(timeoutMs);
		} else {
			std::cout << "Subscriber must be created first!!" << std::endl;
		}
	}

	///Override from ops::DataListener, called whenever new data arrives.
	void onNewData(ops::DataNotifier* subscriber)
	{
		if(subscriber == sub)
		{
			// Check if we have lost any messages. We use the publicationID and that works as long as
			// it is the same publisher sending us messages.
			ops::OPSMessage* newMess = sub->getMessage();

			if (expectedPubId >= 0) {
				if (expectedPubId != newMess->getPublicationID()) {
					std::cout << ">>>>> Lost message for topic " << sub->getTopic().getName() <<
						". Exp.pubid: " << expectedPubId << " got: " << newMess->getPublicationID() << std::endl;
				}
			}
			expectedPubId = newMess->getPublicationID() + 1;
			client->onData(sub, (DataType*)newMess->getData());
		}
	}

	///Override from ops::DeadlineMissedListener, called if no new data has arrived within deadlineQoS.
	void onDeadlineMissed(ops::DeadlineMissedEvent* evt)
	{
		std::cout << "Deadline Missed for topic " << sub->getTopic().getName() << std::endl;
	}
};


std::vector<pizza::special::ExtraAllt> receivedPizzaVec;

typedef CHelper<pizza::special::ExtraAllt, pizza::special::ExtraAlltPublisher, pizza::special::ExtraAlltSubscriber> TExtraAlltHelper;

struct ItemInfo {
	std::string Domain;
	std::string TopicName;
	std::string TypeName;

	bool selected;
	IHelper* helper;
	ops::Participant* part;

	ItemInfo(std::string dom, std::string top, std::string typ)
	{
		Domain = dom;
		TopicName = top;
		TypeName = typ;
		helper = NULL;
		part = NULL;
		selected = false;
	};
};
ItemInfo* itemInfo;

class MyListener :

	public CHelperListener<pizza::special::ExtraAllt>
{
public:

	void onData(ops::Subscriber* sub, pizza::special::ExtraAllt* data){
		receivedPizzaVec.push_back(*data);
		std::cout << "GETS DATA, size = " << receivedPizzaVec.size() << std::endl;
	}
};


//test class
class Test_OPS_Publisher_And_Subscriber : public testing::Test
{
public:
	Test_OPS_Publisher_And_Subscriber() {}

    ~Test_OPS_Publisher_And_Subscriber() {}
};


int main(int argc, char**argv)
{

	// --------------------------------------------------------------------
	MyListener myListener;

	itemInfo = new ItemInfo("PizzaDomain", "ExtraAlltTopic", "pizza.special.ExtraAllt");
	// Setup the OPS static error service (common for all participants, reports errors during participant creation)
	ops::ErrorWriter* errorWriterStatic = new ops::ErrorWriter(std::cout);
	ops::Participant::getStaticErrorService()->addListener(errorWriterStatic);

	// Create participants
	// NOTE that the second parameter (participantID) must be different for the two participant instances
	ops::Participant* participant = ops::Participant::getInstance("PizzaDomain", "PizzaDomain");
	if (participant == NULL) {
	    std::cout << "Failed to create Participant. Missing ops_config.xml ??" << std::endl;
		exit(-1);
	}
	participant->addTypeSupport(new PizzaProject::PizzaProjectTypeFactory());
	ops::Participant* otherParticipant = ops::Participant::getInstance("OtherPizzaDomain", "OtherPizzaDomain");
	if (otherParticipant == NULL) {
		std::cout << "Failed to create Participant. Missing ops_config.xml ??" << std::endl;
        exit(-1);
	}
	otherParticipant->addTypeSupport(new PizzaProject::PizzaProjectTypeFactory());

	// Add error writers to catch internal ops errors
	ops::ErrorWriter* errorWriter = new ops::ErrorWriter(std::cout);
	participant->getErrorService()->addListener(errorWriter);
	ops::ErrorWriter* errorWriter2 = new ops::ErrorWriter(std::cout);
	otherParticipant->getErrorService()->addListener(errorWriter2);

	// Finish up our ItemInfo's
	itemInfo->helper = (new TExtraAlltHelper(&myListener));
	itemInfo->part = participant;
	itemInfo->selected = true;
	itemInfo->helper->CreateSubscriber(itemInfo->part, itemInfo->TopicName);
	itemInfo->helper->StartSubscriber();
	ops::TimeHelper::sleep(22000); //listen on data for 22 seconds

	//delete objects

	delete itemInfo->helper;
	itemInfo->helper = NULL;
	itemInfo->part = NULL;
	delete itemInfo;
	itemInfo = NULL;
	//listen for ten seconds

	participant->getErrorService()->removeListener(errorWriter);
	otherParticipant->getErrorService()->removeListener(errorWriter2);

	delete errorWriter; errorWriter = NULL;
	delete errorWriter2; errorWriter2 = NULL;

	///TODO this should be done by asking Participant to delete instances??
	delete participant;

    //run unit tests
    ::testing::InitGoogleTest(&argc, argv);
    int result = RUN_ALL_TESTS();

    return result;
}


TEST_F(Test_OPS_Publisher_And_Subscriber, TESTING_NORMAL_EXTRA_ALLT) {
	EXPECT_EQ(receivedPizzaVec.size(), 22);
	test::testExtraAlltNormal(receivedPizzaVec.at(0), NORMAL);
}

TEST_F(Test_OPS_Publisher_And_Subscriber, TESTING_LARGE_EXTRA_ALLT) {
	test::testExtraAlltLarge(receivedPizzaVec.at(1), NORMAL);
}

TEST_F(Test_OPS_Publisher_And_Subscriber, TESTING_BURST_NORMAL_EXTRA_ALLT) {
	for(int i = 2; i < 12; ++i) {
		test::testExtraAlltNormal(receivedPizzaVec.at(i), NORMAL);
	}
}

TEST_F(Test_OPS_Publisher_And_Subscriber, TESTING_BURST_LARGE_EXTRA_ALLT) {
	for(int i = 12; i < 22; ++i) {
		test::testExtraAlltLarge(receivedPizzaVec.at(i), NORMAL);
	}
}

File: UnitTests/OPStest-Python/UnitTests/SubscribeDataAndTest.py
import time
import re
import sys
import os
import platform
import unittest


sys.path.append("/home/pane/slask/jali-ops/ops-development/Python")
sys.path.append("/home/pane/slask/pytest/pytest-2.9.1")

from ops import Participant,Publisher,Subscriber,Print_Archiver

curDir = os.getcwd()
sys.path.append(curDir+"/../Pizzas")

import pizza
import pizza_special
import PizzaProjectTypeFactory
#import pytest

#global variables
receivedData = []
delta = 5


#test class normal unit test

class TestExtraAllt(unittest.TestCase):
	def test_normal(self):
		print ""
		for i in range(0,len(receivedData)-1):
			extraAllt = receivedData[i]
			if i == 0 or (i >= 2 and i < 12):
				self.assertEqual(len(extraAllt.bools), 6)
				self.assertEqual(len(extraAllt.bytes), 6)
				self.assertEqual(len(extraAllt.ints) , 5)
				self.assertEqual(len(extraAllt.floats) , 5)
				self.assertEqual(len(extraAllt.doubles) , 5)
				self.assertEqual(len(extraAllt.longs) , 5)
				self.assertEqual(len(extraAllt.strings) , 6)
				self.assertEqual(len(extraAllt.cheeses) , 4)
				
				self.assertEqual( extraAllt.extraCheese 	, True )
				self.assertEqual( extraAllt.nrOfMushRooms	, 14 )
				self.assertEqual( extraAllt.meetQuality	   	, 9 )
				self.assertAlmostEqual( extraAllt.timeBakedHours 	, 123.4, delta )
				self.assertEqual( extraAllt.timeBakedSeconds, 53.4 )
				self.assertEqual( extraAllt.description	   	, "Pizza with extra allt" )
				self.assertEqual( extraAllt.cheese_.age		, 12.0 )
				self.assertEqual( extraAllt.cheese_.name	, "gorgonzola" )
				
				
				self.assertEqual( extraAllt.bools   		, [True, False, True, False, True, False] )
				self.assertEqual( extraAllt.bytes   		, [-64, -32, -16, 15, 31, 63] )
				self.assertEqual( extraAllt.ints    		, [0, 123, -523, 1000, -5000] )
				self.assertEqual( extraAllt.longs   		, [0, 123, -523, 1000, -5000] )
				self.assertEqual( extraAllt.floats  		, [0.0, 123.0, -523.0, 1000.0, -5000.0] )
				self.assertEqual( extraAllt.doubles 		, [0.0, 123.0, -523.0, 1000.0, -5000.0] )
				self.assertEqual( extraAllt.strings 		, ["extra", "allt", "er", "den", "basta", "pizzan"] )
				self.assertEqual( extraAllt.cheeses[0].age 	, 1 )
				self.assertEqual( extraAllt.cheeses[0].name , "ost1" )
				self.assertEqual( extraAllt.cheeses[1].age 	, 2 )
				self.assertEqual( extraAllt.cheeses[1].name , "ost2" )
				self.assertEqual( extraAllt.cheeses[2].age 	, 3 )
				self.assertEqual( extraAllt.cheeses[2].name , "ost3" )
				self.assertEqual( extraAllt.cheeses[3].age 	, 4 )
				self.assertEqual( extraAllt.cheeses[3].name , "ost4" )
			else:
				self.assertEqual( extraAllt.extraCheese 	, True )
				self.assertEqual( extraAllt.nrOfMushRooms	, 14 )
				self.assertEqual( extraAllt.meetQuality	   	, 9 )
				self.assertAlmostEqual( extraAllt.timeBakedHours 	, 123.4, delta )
				self.assertEqual( extraAllt.timeBakedSeconds, 53.4 )
				self.assertEqual( extraAllt.description	   	, "Pizza with extra allt" )
				self.assertEqual( extraAllt.cheese_.age		, 12.0 )
				self.assertEqual( extraAllt.cheese_.name	, "gorgonzola" )
				
				self.assertEqual(len(extraAllt.bools), 200)
				self.assertEqual(len(extraAllt.bytes), 200)
				self.assertEqual(len(extraAllt.ints) , 200)
				self.assertEqual(len(extraAllt.floats) , 200)
				self.assertEqual(len(extraAllt.doubles) , 200)
				self.assertEqual(len(extraAllt.longs) , 200)
				self.assertEqual(len(extraAllt.strings) , 200)
				self.assertEqual(len(extraAllt.cheeses) , 200)
				for j in range(0,200):
					if j <=	 100:
						self.assertEqual(extraAllt.bools[j] , True )
						self.assertEqual(extraAllt.bytes[j] , -42 )
						self.assertEqual(extraAllt.ints[j] , 5 )
						self.assertEqual(extraAllt.floats[j] , 15.0 )
						self.assertEqual(extraAllt.doubles[j] , 25.0 )
						self.assertEqual(extraAllt.longs[j] , 35 )
						self.assertEqual(extraAllt.strings[j] , "hejsan" )
						self.assertEqual(extraAllt.cheeses[j].age , 3 )
						self.assertEqual(extraAllt.cheeses[j].name , "ecklig ost" )
					else:
						self.assertEqual(extraAllt.bools[j] , False )
						self.assertEqual(extraAllt.bytes[j] , 42 )
						self.assertEqual(extraAllt.ints[j] , 10 )
						self.assertEqual(extraAllt.floats[j] , 20.0 )
						self.assertEqual(extraAllt.doubles[j] , 30.0 )
						self.assertEqual(extraAllt.longs[j] , 40 )
						self.assertEqual(extraAllt.strings[j] , "hoppsan" )
						self.assertEqual(extraAllt.cheeses[j].age , 6 )
						self.assertEqual(extraAllt.cheeses[j].name , "god ost" )


										
					

def onExtraAllt(sub,mess):
	addr,port = map(str,mess.getSource())
	data=mess.data
	receivedData.append(data)



class CHelper(object):
	def __init__(self):
		super(CHelper,self).__init__()
		self.callback = None
		self.pub = None
		self.sub = None
		self.expectedPubId = -1
		self.data = None


	def CreateSubscriber(self,part,topicName):
		if self.sub is not None:
			print "Subscriber already exist for topic " + self.sub.getTopic().getName()
		else:
			topic = part.createTopic(topicName);
			print "Created topic " + topic.name + " [" + topic.transport + "." + topic.domainAddress +"." + str(topic.port) + "]"
			self.sub = Subscriber.Subscriber(topic)
			#self.sub.addCallback(self.listener.onNewData)
			if self.callback is not None:
				self.sub.addCallback(self.callback)
			self.sub.start()
			
	def DeleteSubscriber(self,doLog = True):
		print "Deleting subscriber for topic " + self.sub.getTopic().getName()
		self.sub.stop();
		self.sub = None;

	def StartSubscriber(self):
		self.sub.start();
	
	def StopSubscriber(self):
		self.sub.stop()


class ItemInfo(object):
	def __init__(self,dom,top,typ):
		self.Domain = dom;
		self.TopicName = top
		self.TypeName = typ
		self.helper = None
		self.part = None
		self.selected = False
	def __str__(self):
		return ("P" if self.helper.HasPublisher() else " ") + ("S" if self.helper.HasPublisher() else " ") + ("*" if self.selected else " ") + self.Domain + "::" + self.TopicName

ItemInfoList = []

ItemInfoList.append(ItemInfo("PizzaDomain", "ExtraAlltTopic", "pizza.special.ExtraAllt"))


participant = Participant.Participant.getInstance("PizzaDomain", "PizzaDomain", "../ops_config.xml")
if participant == None:
	print "Failed to create Participant. Missing ops_config.xml ??"
	sys.exit(-1)
participant.addTypeSupport(PizzaProjectTypeFactory.PizzaProjectTypeFactory())


info = ItemInfoList[0]

#create subsrcriber
info.helper = CHelper()
info.helper.data = pizza_special.ExtraAllt()
info.helper.callback = onExtraAllt
info.part = participant

info.helper.CreateSubscriber(info.part, info.TopicName)
info.helper.StartSubscriber()

#wait until data is collected
time.sleep(11)

#delete subscriber
#info.helper.StopSubscriber()
#info.helper.DeleteSubscriber()


if __name__ == '__main__':
	suite = unittest.TestLoader().loadTestsFromTestCase(TestExtraAllt)
	unittest.TextTestRunner(verbosity=2).run(suite)



